### ISSUE TITLE
Clarification on Segmentation Results ? 

### ISSUE THREAD

**Issue by ching-sui1995** [MzeroMiko/VMamba#328]
Can you share more details about your segmentation experiments ? 

I am interested in VMamba-T[s1l8] for example. The config [here](https://github.com/MzeroMiko/VMamba/blob/main/segmentation/configs/vssm1/upernet_vssm_4xb4-160k_ade20k-512x512_tiny.py) says nothing about the learning rate, batch size, etc. as one may assume it is the same as Swin-T [config](https://github.com/MzeroMiko/VMamba/blob/main/segmentation/configs/swin/swin-tiny-patch4-window7-in1k-pre_upernet_8xb2-160k_ade20k-512x512.py). 

But this just makes no sense as VMamba-T significantly outperforms Swin even from the first 16k iterations. 

See below for example:

Here's the performance of Swin-T:
mIoU: 0.3591
{"mode": "val", "epoch": 13, "iter": 16000, "lr": 5e-05, "mIoU": 0.3591, "mAcc": 0.4795, "aAcc": 0.778}

Here's the performance of VMamba-T at the same 16000 iterations:
mIoU: 42.0500  mAcc: 54.8000 

How can it be almost +6.14 mIoU better in the first 16000 ?

If you use the same exact training procedure, then performance should be comparable ?

**Comment by ching-sui1995** [comment]
I found another confusing issue about the segmentation results. So according to this, if you change the validation batch size then results are different #218 ?  

This should not be the case !

Can you comment what is happening ?

**Comment by ching-sui1995** [comment]
Just for the record, here's the progression of VMamba-T in terms of mIoU for segmentation task:

Iteration (k)      mIoU
16                    42.05
16                    42.05
16                    42.05
16                    42.05

**Comment by MzeroMiko** [comment]
Here's the full log (and there's also full config inside) https://github.com/MzeroMiko/VMamba/releases/download/%23v2seg/upernet_vssm_4xb4-160k_ade20k-512x512_tiny.log

You may find more details in `releases` page.

**Comment by ching-sui1995** [comment]
But this does not explain why changing batch size at test/validation time leads to such discrepancy as reported her #218 ? 

Shouldn't the validation results be the same if I use batch size 1 or 2 ?

**Comment by MzeroMiko** [comment]
I think that is the difference of batchsize in training, as I assume that one may use the same batchsize in val as in train (actually, one may not even change the config, the batchsize difference comes from the number of gpus he/she uses). 

Changing valset batchsize do affect the performance if taking the precision of implementation into account, but the difference may just in 1e-3 or even smaller.

**Comment by ching-sui1995** [comment]
Regarding batch size, I believe it is the eval batch size which is different (250 vs 500) but should not result in such differences. 

But in general, why do you think your model starts off with such a good mIoU accuracy (42.05) after only 16k iterations of training ? 

This is not seen any other tiny-sized pretrained ImageNet models.

**Comment by MzeroMiko** [comment]
To clarify, I agree that eval batch size do not affect the performance much, and I think in that issue #218 it's the training batch size difference that make the performance change. 

I think the performance improvement may comes from the unique design of cross-scan, together with the characteristic of RNN-based models. The former inject the 2D prior and instance-prior into the network, while the latter collect the information far from that patch.  And that's what CNN or transformer can not learn in just few steps.

**Comment by ching-sui1995** [comment]
Just to clarify, you used these settings in all segmentation experiments ? 

Total batch size 16

And the following ? 

```
optim_wrapper = dict(
    _delete_=True,
    type='OptimWrapper',
    optimizer=dict(
        type='AdamW', lr=0.00006, betas=(0.9, 0.999), weight_decay=0.01),
    paramwise_cfg=dict(
        custom_keys={
            'absolute_pos_embed': dict(decay_mult=0.),
            'relative_position_bias_table': dict(decay_mult=0.),
            'norm': dict(decay_mult=0.)
        }))

param_scheduler = [
    dict(
        type='LinearLR', start_factor=1e-6, by_epoch=False, begin=0, end=1500),
    dict(
        type='PolyLR',
        eta_min=0.0,
        power=1.0,
        begin=1500,
        end=160000,
        by_epoch=False,
    )
]
```

Maybe cross-scan is effective, but still not confident if such changes can lead to massive performance improvement, especially that Mamba itself is a linear model.

I am just trying to figure out if better performance is due to better training procedure used by your work.

**Comment by MzeroMiko** [comment]
I used the config file exactly as the shown in this repo, which, if I did not write it wrong, is the same config (except for the model) as in swin. 

I even used basically the same configuration as in swin in classification task, only changing the architecture it self and the drop-path-rate if needed. 

So it is very likely that if you change the configuration, you may get better performance.

**Comment by MzeroMiko** [comment]
My experiments support your statement. But it mainly dues to the relatively low efficiency of the practical implementation of parallel scan operation compared to matrix multiplication in modern GPUs. 

And in the paper, you may find more interesting characteristics about VMamba.

![image](https://github.com/user-attachments/assets/1a0675ef-a09d-4e65-8a83-f82688049eda)

**Comment by ching-sui1995** [comment]
After spending more time with VMamba, I am feeling it is somehow over-fitting in most downstream tasks. This is because I tried to increase drop_path or decay and results got better. But I am just not sure if your model can actually scale up because when data becomes more available, you cannot overfit the way it does on ADE or COCO. Or at least this is what I saw, maybe I am wrong. Good job any ways !
